codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -250,79 +251,168 @@ private void writePendingBytes() throws Http2Exception {
-        // upon stream weight relative to its peers
-        Http2Stream[] children = parent.children().toArray(new Http2Stream[parent.numChildren()]);
-        int totalWeight = parent.totalChildWeights();
-        for (int tail = children.length; tail > 0;) {
-            int head = 0;
-            int nextTail = 0;
-            int nextTotalWeight = 0;
-            int nextConnectionWindow = connectionWindow;
-            for (; head < tail && nextConnectionWindow > 0; ++head) {
-                Http2Stream child = children[head];
-                state = state(child);
-                int weight = child.weight();
-                double weightRatio = weight / (double) totalWeight;
-
-                // In order to make progress toward the connection window due to possible rounding errors, we make sure
-                // that each stream (with data to send) is given at least 1 byte toward the connection window.
-                int connectionWindowChunk = Math.max(1, (int) (connectionWindow * weightRatio));
-                int bytesForTree = Math.min(nextConnectionWindow, connectionWindowChunk);
-                int bytesForChild = Math.min(state.streamableBytes(), bytesForTree);
-
-                if (bytesForChild > 0) {
-                    state.allocate(bytesForChild);
-                    bytesAllocated += bytesForChild;
-                    nextConnectionWindow -= bytesForChild;
-                    bytesForTree -= bytesForChild;
-                    // If this subtree still wants to send then re-insert into children list and re-consider for next
-                    // iteration. This is needed because we don't yet know if all the peers will be able to use
-                    // all of their "fair share" of the connection window, and if they don't use it then we should
-                    // divide their unused shared up for the peers who still want to send.
-                    if (state.streamableBytesForTree() > 0) {
-                        children[nextTail++] = child;
-                        nextTotalWeight += weight;
+
+            return nextConnectionWindow > 0;
+        }
+
+        void feedHungryChildren() throws Http2Exception {
+            if (stillHungry == null) {
+                // There are no hungry children to feed.
+                return;
+            }
+
+            totalWeight = nextTotalWeight;
+            connectionWindow = nextConnectionWindow;
+
+            // Loop until there are not bytes left to stream or the connection window has collapsed.
+            for (int tail = nextTail; tail > 0 && connectionWindow > 0;) {
+                nextTotalWeight = 0;
+                nextTail = 0;
+
+                // Iterate over the children that are currently still hungry.
+                for (int head = 0; head < tail && nextConnectionWindow > 0; ++head) {
+                    Http2Stream child = stillHungry[head];
+                    if (!visit(child)) {
```

Comment:
Inline the `child` variable? it is only used 1 time...less load/store operations. `if (!visit(stillHungry[head])) {`